File: ai/linear_model.py
#!/usr/bin/env python3
"""Simple linear model from command-line arguments."""

import sys
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

PLOT_NAME = "regression_plot_python.png"


def script_dir() -> Path:
    return Path(__file__).resolve().parent


def usage() -> None:
    print(f"Usage: {Path(sys.argv[0]).name} <filename> <x_column> <y_column>")
    sys.exit(1)


def main() -> None:
    if len(sys.argv) != 4:
        usage()

    filename, x_column, y_column = sys.argv[1:4]
    data_path = Path(filename)
    if not data_path.is_file():
        print(f"Error: file not found: {filename}")
        sys.exit(1)

    data = pd.read_csv(data_path)
    if x_column not in data.columns or y_column not in data.columns:
        print(f"Error: columns must exist in {filename}")
        print(f"Available columns: {', '.join(data.columns)}")
        sys.exit(1)

    x = data[[x_column]].values
    y = data[y_column].values

    model = LinearRegression()
    model.fit(x, y)
    y_pred = model.predict(x)

    slope = float(model.coef_[0])
    intercept = float(model.intercept_)
    pearson_r = float(np.corrcoef(data[x_column], data[y_column])[0, 1])
    mse = float(mean_squared_error(y, y_pred))
    r2 = float(r2_score(y, y_pred))

    print("Python Linear Model Results")
    print("===========================")
    print(f"File: {filename}")
    print(f"X column: {x_column}")
    print(f"Y column: {y_column}")
    print(f"Observations: {len(data)}")
    print(f"Slope ({x_column}): {slope:.4f}")
    print(f"Intercept: {intercept:.4f}")
    print(f"Pearson's r: {pearson_r:.6f}")
    print(f"MSE: {mse:.4f}")
    print(f"R-squared (optional): {r2:.6f}")

    fig, ax = plt.subplots(figsize=(9, 6))
    ax.scatter(data[x_column], data[y_column], color="steelblue", label="Observed")

    x_line = np.linspace(data[x_column].min(), data[x_column].max(), 100)
    ax.plot(x_line, intercept + slope * x_line, color="crimson", linewidth=2, label="Regression line")

    equation = f"{y_column} = {intercept:,.0f} + {slope:,.0f} * {x_column}"
    annotation = (
        f"{equation}\n"
        f"Pearson's r = {pearson_r:.4f}\n"
        f"MSE = {mse:,.0f}\n"
        f"R² = {r2:.4f}"
    )
    ax.text(
        0.05,
        0.95,
        annotation,
        transform=ax.transAxes,
        va="top",
        ha="left",
        bbox=dict(boxstyle="round,pad=0.5", facecolor="white", edgecolor="gray", alpha=0.9),
        fontsize=10,
    )

    ax.set_xlabel(x_column)
    ax.set_ylabel(y_column)
    ax.set_title(f"{y_column} vs {x_column} with Annotated Regression Line")
    ax.legend(loc="lower right")
    fig.tight_layout()

    plot_path = script_dir() / PLOT_NAME
    fig.savefig(plot_path, dpi=150)
    plt.close(fig)
    print(f"Plot saved to {plot_path}")


if __name__ == "__main__":
    main()

File: ai/linear_model.R
#!/usr/bin/env Rscript
# Simple linear model from command-line arguments.

args <- commandArgs(trailingOnly = TRUE)
args_all <- commandArgs(trailingOnly = FALSE)
script_path <- sub("^--file=", "", args_all[grep("^--file=", args_all)])
script_dir <- if (length(script_path)) dirname(normalizePath(script_path)) else getwd()
plot_file <- file.path(script_dir, "regression_plot_r.png")

usage <- function() {
  script <- basename(commandArgs(trailingOnly = FALSE)[grep("^--file=", commandArgs(trailingOnly = FALSE))])
  script <- if (length(script)) sub("^--file=", "", script) else "linear_model.R"
  cat(sprintf("Usage: %s <filename> <x_column> <y_column>\n", script))
  quit(status = 1)
}

if (length(args) != 3) {
  usage()
}

filename <- args[1]
x_column <- args[2]
y_column <- args[3]

if (!file.exists(filename)) {
  cat(sprintf("Error: file not found: %s\n", filename))
  quit(status = 1)
}

data <- read.csv(filename, stringsAsFactors = FALSE)
if (!(x_column %in% names(data)) || !(y_column %in% names(data))) {
  cat(sprintf("Error: columns must exist in %s\n", filename))
  cat(sprintf("Available columns: %s\n", paste(names(data), collapse = ", ")))
  quit(status = 1)
}

formula <- as.formula(paste(y_column, "~", x_column))
model <- lm(formula, data = data)

coefs <- coef(model)
intercept <- unname(coefs[1])
slope <- unname(coefs[2])
pearson_r <- cor(data[[x_column]], data[[y_column]])
mse <- mean(residuals(model)^2)
r2 <- summary(model)$r.squared

cat("R Linear Model Results\n")
cat("========================\n")
cat(sprintf("File: %s\n", filename))
cat(sprintf("X column: %s\n", x_column))
cat(sprintf("Y column: %s\n", y_column))
cat(sprintf("Observations: %d\n", nrow(data)))
cat(sprintf("Slope (%s): %.4f\n", x_column, slope))
cat(sprintf("Intercept: %.4f\n", intercept))
cat(sprintf("Pearson's r: %.6f\n", pearson_r))
cat(sprintf("MSE: %.4f\n", mse))
cat(sprintf("R-squared (optional): %.6f\n", r2))

x_vals <- data[[x_column]]
y_vals <- data[[y_column]]
x_label <- if (x_column == "YearsExperience") "Years of Experience" else x_column
plot_title <- paste(y_column, "vs", x_label, "with Annotated Regression Line")

equation <- sprintf(
  "%s = %s + %s * %s\nPearson's r = %.4f\nMSE = %s\nR² = %.4f",
  y_column,
  format(round(intercept), big.mark = ",", scientific = FALSE),
  format(round(slope), big.mark = ",", scientific = FALSE),
  x_column,
  pearson_r,
  format(round(mse), big.mark = ",", scientific = FALSE),
  r2
)

library(ggplot2)

x_range <- max(x_vals) - min(x_vals)
label_x <- if (x_range > 0) min(x_vals) + 0.05 * x_range else min(x_vals)
label_y <- max(y_vals)

plot <- ggplot(data, aes(x = .data[[x_column]], y = .data[[y_column]])) +
  geom_point(color = "steelblue", size = 3) +
  geom_abline(
    intercept = intercept,
    slope = slope,
    color = "red",
    linewidth = 1
  ) +
  annotate(
    "label",
    x = label_x,
    y = label_y,
    label = equation,
    hjust = 0,
    vjust = 1,
    size = 3.5,
    linewidth = 0.2
  ) +
  labs(
    title = plot_title,
    x = x_label,
    y = y_column
  )

ggsave(plot_file, plot = plot, width = 9, height = 6, dpi = 150)

cat(sprintf("Plot saved to %s\n", plot_file))
